### DISCUSSION TITLE
In the preprocessor Add a function to create  a JSON, add and remove JSON properties

### DISCUSSION THREAD

**Discussion by JohnLeyva** [plantuml/plantuml#1733] [Ideas]
The proposal is to add the following functions to the proprocesor

%json(jsonstring)
%json_set(jsonvar, propname, value)
%json_remove(jsonvar, propname)

The following provides some examples how it can be used.

``` plantuml
card jsonExample [
    
    ' Creates a JSON from a string
    '!$r1 = %json('{ "a" : 1, "b" : "two" }')
    !$r1 = { "a" : 1, "b" : "two" }
    r1 = $r1

    ' Set a JSON child 
    '!$r2 = %json_set($r1,'c',3)
    !$r2 = { "a" : 1, "b" : "two", "c" : 3 } 
    r2 = $r2

    ' Set a JSON sub child 
    '!$r3_d = %json($r1,'{ "da" : 1, "db" : "two" }')
    !$r3_d = { "da" : 1, "db" : "two" }
    '!$r3 = %json_set($r1,"d",$r3_d)
    '   The following does not work at the moment
    '   !$r3 = { "a" : 1, "b" : "two", "d" : $r3_d }
    !$r3 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two" } }
    r3 = $r3

    !$r4_d_c = { "a" : [1,2] }
    '!$r4 = %json_set($r3.d,'dc',r3_d_c.a)
    !$r4 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two", "dc" : [1,2] } }
    r4 = $r4

    '!$r1 = %json_remove($r4,'d')
    !$r1 = { "a" : 1, "b" : "two" }
    r1 = $r1

    '!$r3_d = %json_remove($r4.d,'dc')
    !$r3_d = { "a" : 1, "b" : "two" }
    !$r3 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two" } }
    
    r3_d = $r3_d
    r3 = $r3
]
```

@The-Lum, I see you recentlu worked on the proprocessor so perhaps it is something of your interest

**Comment by The-Lum** [comment]
Hello @JohnLeyva, and all,

Just FYI, here are some pseudo-similar requests here:

- https://forum.plantuml.net/15436/allow-update-json-object-or-array-value
- https://forum.plantuml.net/17474/modify-json-variables-in-preprocessor
- https://forum.plantuml.net/18418/request-%25set_json_keys-function-json-writing-capabilities
- https://github.com/plantuml/plantuml/issues/328

I actually worked on the preprocessor...
And I can try to look on that but I can't promise anything...
_Step by step..._

Regards,
Th.

**Reply by The-Lum** [reply]
Hello @JohnLeyva, and all,

Here is a fisrt proposal _(I propose to change a little the function name...)_ with:
- `%str2json('json_string')`
- `%json_add(jsonvar, propname,  jsonvar)` or  `%json_add(jsonvar, jsonvar_to_add)`
- `%json_remove(propname)` or `%json_remove(index)`

Because it will be necessary to considering all JSON structures (Array, Object, Number, String, ...)...

---
And the fisrt tests _(of the day...)_:

## `%str2json`
```puml
%str2json('{"a": 1, "b": "two"}') = {"a":1,"b":"two"}
```
## `%json_add`
### On Object Context
```puml
%json_add({"a": 1, "b": "two"}, c, 3) = {"a":1,"b":"two","c":3} 
%json_add({"a": 1, "b": "two"}, d, {"da": 1, "db": "two"}) = {"a":1,"b":"two","d":{"da":1,"db":"two"}}
```
### On Array Context
```puml
%json_add([0], "a") = [0,"a"]
```
## `%json_remove`
### On Object Context
```puml
%json_remove({"a": 1, "b": "two","c":3}, c) = {"a":1,"b":"two"}
%json_remove({"a":1,"b":"two","d":{"da":1,"db":"two"}}, d) = {"a":1,"b":"two"}
```
### On Array Context
```puml
%json_remove([1,2,3], 0) = [2,3]
```
- What should we do when the index is negative? _(nothing or from the end of the array.. ??)_
---

- What do you think?

For debug purpose, it will be perhaps necessary to implement a print function (as `%json_print`, or `%json2str`)...
- What will be the better name?

_Remains some bondaries and side effects checks... in order to commit and PR..._

_See you for next step..._
Regards,
Th.

**Reply by The-Lum** [reply]
Hello everybody,

Here is a first dev. attempt with:
- #1742

Available on last [snapshot (~v1.2024.5beta1)](https://github.com/plantuml/plantuml/releases/tag/snapshot).

- _Could you test to confirm proper operation?_

_Enjoy,_
Regards,
Th.

**Reply by JohnLeyva** [reply]
Thanks @The-Lum 

Just tested it with the following

```
@startuml


card jsonExample [
    %version()

    
    ' Creates a JSON from a string
    !$r1 = %str2json('{ "a" : 1, "b" : "two" }')
    '!$r1 = { "a" : 1, "b" : "two" }
    r1 = $r1

    ' Set a JSON child 
    !$r2 = %json_add($r1,'c',3)
    '!$r2 = { "a" : 1, "b" : "two", "c" : 3 } 
    r2 = $r2

    ' Set a JSON sub child 
    !$r3_d = %str2json('{ "da" : 1, "db" : "two" }')
    '!$r3_d = { "da" : 1, "db" : "two" }
    !$r3 = %json_add($r1,"d",$r3_d)
    !$r3 = %json_add($r3,"e",$r3_d)
    '   The following does not work at the moment
    '!$r3 = { "a" : 1, "b" : "two", "d" : $r3_d }
''    !$r3 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two" } }
    r3 = $r3

    !$r4_d_c = { "a" : [1,2] }
    '!$r4 = %json_add($r3.d,'dc',r3_d_c.a)
    !$r4 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two", "dc" : [1,2] } }
    r4 = $r4

    !$r1 = %json_remove($r4,'d')
    '!$r1 = { "a" : 1, "b" : "two" }
    r1 = $r1

    !$r3_d = %json_remove($r4.d,'dc')
    '!$r3_d = { "a" : 1, "b" : "two" }
    !$r3 = { "a" : 1, "b" : "two", "d" : { "da" : 1, "db" : "two" } }
    
    r3_d = $r3_d
    r3 = $r3
]

@enduml
```

**Reply by The-Lum** [reply]
Hi @JohnLeyva,

See also new JSON function: `%json_set`, `%json_merge`,... and doc. here:
- https://plantuml.com/preprocessing-json

_Enjoy,_
Regards,
Th.

**Comment by matthew16550** [comment]
https://jsonpatch.com/ might provide some ideas and the RFC has a list of examples that could be helpful test cases.

Combining with @The-Lum  proposal, perhaps:

`%json_add(jsonvar, path,  value)`
`%json_remove(jsonvar, path)`

JSON Patch uses [JSON Pointer](https://datatracker.ietf.org/doc/html/rfc6901/) to describe the locations but that is a different format than plantuml has been using so more thought needed.
